### PR TITLE
feat: add quality-review feature with weighted FQI rubric and Customer Lens

### PR DESCRIPTION
**Pull request by gvauter**
## Summary

Adds a new `quality-review` feature that scores CVE flaw content against the FQI (Flaw Quality Index) rubric, evaluated through a Customer Lens framework.

## What it does

- Scores 30 criteria across 6 categories (5 criteria x 0-2 points each)
- Applies FQI category weights: Description 20%, Statement 25%, Mitigation 10%, Grammar 15%, Content Ambiguity 15%, Technical Value 15%
- Produces a weighted overall score on a **0.0-1.0 scale** (consistent with `confidence` and `data_quality` in `AegisFeatureModel`) with quality tiers (Excellent >=0.8, Good >=0.6, Needs Improvement >=0.4, Fails Standards <0.4)
- Evaluates content through 3 customer personas (Ops/Sysadmin, Security/CISO, Compliance/Auditor)
- Generates suggested statement/mitigation rewrites when scores are low
- Criterion IDs and scoring definitions match `rubric.json` from the FQI dashboard exactly

## Changes

- **`data_models.py`**: New models — `QualityReviewModel`, `CriterionScore`, `CustomerLensAssessment`, `QualityRating`, `CATEGORY_WEIGHTS`. Auto-computed weighted score with clamping and 0.0-1.0 normalization.
- **`cve/__init__.py`**: `QualityReview` feature class with full rubric prompt (Score=2/1/0 definitions, evidence fields, LLM guidance) and `_check_output` validation enforcing all 30 criteria with unique IDs across 6 categories.
- **`main.py` (CLI)**: `aegis quality-review <CVE-ID>` command
- **`main.py` (web)**: Registered as `quality-review` endpoint
- **Tests**: 19 unit tests (model validation, weighted scoring, tier boundaries, weight verification) + 1 integration test with strengthened assertions (rubric shape, criterion uniqueness, disclaimer, customer lens)

## Reviewer feedback addressed (from PR #684)

- **Score range**: Normalized to 0.0-1.0 (was 0-10) per maintainer request for consistency with `AegisFeatureModel`
- **Category name duplication**: `_REQUIRED_CATEGORIES` now derived from `CATEGORY_WEIGHTS.keys()`
- **Raw score clamping**: Categories clamped to max 10 raw points
- **Output validation**: `_check_output` enforces exactly 30 scores with unique criterion IDs
- **Integration test**: Validates rubric shape, category set, criterion uniqueness, disclaimer text, customer lens population

Supersedes #684 (reopened from upstream branch for CI access).

## Summary by Sourcery

Introduce a new CVE quality-review feature that evaluates flaw content against a weighted rubric and exposes it via CLI, web API, and supporting data models.

New Features:
- Add QualityReview feature for CVEs that scores flaw content using a 30-criterion, 6-category FQI rubric and Customer Lens framework.
- Expose quality-review as a CLI command for assessing a specific CVE ID.
- Register quality-review as a web endpoint for remote evaluation of CVE content.

Enhancements:
- Define data models for quality review scoring, customer lens assessment, and overall quality rating with auto-computed weighted scores on a 0.0-1.0 scale.
- Enforce strict output validation for the QualityReview feature to ensure all rubric criteria and categories are present and uniquely identified.

Tests:
- Add unit tests covering quality review model validation, weighting logic, rating boundaries, and score auto-computation behavior.
- Add an integration test for the QualityReview feature that validates rubric shape, category coverage, criterion uniqueness, disclaimer text, and customer lens population.

### CODE REVIEW

File: src/aegis_ai/features/cve/__init__.py [hunk 1]
```diff
@@ -799,3 +801,363 @@ async def exec(self, cve_id: CVEID, static_context: Any = None):
...
+    _REQUIRED_CRITERIA: set[str] = {
+        # Description — Technical Clarity
+        "component_location",
+        "vuln_type_mechanics",
+        "trigger_conditions",
+        "impact_sketch_cia",
+        "clarity_for_non_experts",
+        # Statement — Technical Clarity
+        "presence_per_cvss_rule",
+        "structured_formula_present",
```
[the remark this answers is not in the document]
Comment (gvauter) [reply src/aegis_ai/features/cve/__init__.py:820]: Addressed in 39cf768 — `_check_output` now uses `len(self._REQUIRED_CRITERIA)` throughout instead of hardcoded `30`. Also added rejection of unexpected category names in the same commit.

File: tests/test_quality_review_models.py [hunk 2]
```diff
@@ -0,0 +1,275 @@
+import pytest
+
+from pydantic import ValidationError
```
[the remark this answers is not in the document]
Comment (gvauter) [reply tests/test_quality_review_models.py:3]: Addressed in 39cf768 — added `test_category_raw_scores_clamped_to_10` which constructs 6 criteria (12 raw points) for the first category and verifies the score is clamped to the category weight (as if 10/10).

File: tests/test_features.py [hunk 3]
```diff
@@ -139,6 +140,60 @@ async def test_component_intelligence_test_model():
...
+async def test_quality_review_with_test_model():
+    test_name = "test_quality_review_with_test_model"
+
+    result = get_cached_response(test_name)
+
+    if not result:
+        llm_result = await cve.QualityReview(rh_feature_agent).exec("CVE-2025-0725")
+        result = llm_result.output.model_dump_json(indent=4)
+        cache_response(test_name, result)
+
```
[the remark this answers is not in the document]
Comment (gvauter) [reply tests/test_features.py:153]: Skipping this one — `_check_output` is a guardrail hook called internally by `guarded_run` with a `(result, deps)` signature tied to the pydantic-ai run lifecycle. Unit testing it in isolation would require mocking that interface, and the value is low since the integration test already exercises the valid path, and the guardrail's job is to trigger LLM retries (not raise exceptions). The validation logic itself is straightforward set math.

File: tests/test_features.py [hunk 4]
```diff
@@ -139,6 +140,60 @@ async def test_component_intelligence_test_model():
...
+        f"Expected 30 criterion scores, got {len(quality_review.scores)}"
+    )
+    categories = {s.category for s in quality_review.scores}
+    assert categories == set(CATEGORY_WEIGHTS.keys()), (
+        f"Categories mismatch: {categories}"
+    )
+
+    # Verify all 30 criterion IDs are unique
+    criterion_ids = [s.criterion_id for s in quality_review.scores]
+    assert len(set(criterion_ids)) == 30, (
```
[the remark this answers is not in the document]
Comment (gvauter) [reply tests/test_features.py]: Addressed in 39cf768 — integration test now asserts `criterion_ids == QualityReview._REQUIRED_CRITERIA` for an exact set match, not just uniqueness count.

File: src/aegis_ai/features/cve/data_models.py [hunk 5]
```diff
@@ -346,3 +347,170 @@ class CVSSDiffExplainerModel(AegisFeatureModel):
...
+        ...,
+        description="What REMAINS UNCLEAR from the current content.",
+    )
+    manual_context_needed: List[str] = Field(
+        ...,
+        description="What additional context an analyst would need to add MANUALLY.",
+    )
+
+
+class QualityReviewModel(AegisFeatureModel):
```
Comment (kdudka) [inline src/aegis_ai/features/cve/data_models.py:419]: `QualityReviewModel` is the only `AegisFeatureModel` subclass that does not include the `explanation` field. All other 10 subclasses have it.  I suspect this discrepancy may surprise the consumers of Aegis suggestions which expect the `explanation` field to be always available.

File: src/aegis_ai/features/cve/data_models.py [hunk 6]
```diff
@@ -346,3 +347,170 @@ class CVSSDiffExplainerModel(AegisFeatureModel):
...
+        ...,
+        description="Critical gaps that must be addressed.",
+    )
+
+    recommendations: List[str] = Field(
+        ...,
+        description="Actionable recommendations to improve the content.",
+    )
+
+    suggested_statement: Optional[str] = Field(
```
Comment (kdudka) [inline src/aegis_ai/features/cve/data_models.py:469]: When the `suggested_statement` and `suggested_mitigation` fields are not `null`, they sometimes violate the rules that we enforce (and evaluate) for the `suggest-statement` feature, such as ["do not suggest update as a mitigation"](https://github.com/RedHatProductSecurity/aegis-ai/blob/4dd13a56fa9862a9ac506bfe8183a5abf70359fb/src/aegis_ai/features/cve/__init__.py#L695), or ["statement should not duplicate description"](https://github.com/RedHatProductSecurity/aegis-ai/blob/4dd13a56fa9862a9ac506bfe8183a5abf70359fb/evals/features/cve/test_suggest_statement.py#L520).

For example `uv run aegis quality-review CVE-2026-4724` suggests the following mitigation, which does not sound appropriate for a Red Hat product:
> Users should update to Firefox version 149 or later, and Firefox ESR version 140.9 or later, to remediate this vulnerability. Refer to the Mozilla security advisories (mfsa2026-22, mfsa2026-24) for further details.

File: src/aegis_ai/features/cve/__init__.py [hunk 7]
```diff
@@ -799,3 +801,370 @@ async def exec(self, cve_id: CVEID, static_context: Any = None):
...
+        return None
+
+    async def exec(self, cve_id: CVEID, static_context: Any = None):
+        """Run the quality review rubric against the given CVE flaw content."""
+        deps = feature_deps(
+            exclude_osidb_fields=[],
+            static_context=static_context,
+        )
+
+        prompt = AegisPrompt(
```
Comment (kdudka) [inline src/aegis_ai/features/cve/__init__.py:900]: I am not sure if it is caused by the `prompt`, the output data model, or their combination, but this feature seems to trigger the following warnings on many CVEs:
```
2026-06-17 16:52:41 [WARNING] QualityReview(CVE-2023-48795): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:52:43 [WARNING] QualityReview(CVE-2023-53188): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:52:48 [WARNING] QualityReview(CVE-2022-50390): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:53:51 [WARNING] QualityReview(CVE-2025-15284): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:54:26 [WARNING] QualityReview(CVE-2025-39809): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:54:30 [WARNING] QualityReview(CVE-2025-39816): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:54:37 [WARNING] QualityReview(CVE-2025-39822): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:54:44 [WARNING] QualityReview(CVE-2025-39809): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:54:46 [WARNING] QualityReview(CVE-2025-39816): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:54:51 [WARNING] QualityReview(CVE-2025-39822): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:03 [WARNING] QualityReview(CVE-2025-53020): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:18 [WARNING] QualityReview(CVE-2025-53020): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:27 [WARNING] QualityReview(CVE-2025-66623): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:34 [WARNING] QualityReview(CVE-2025-53020): output enforcement retry 3/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:39 [WARNING] QualityReview(CVE-2026-5483): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:44 [WARNING] QualityReview(CVE-2025-53020): output enforcement exhausted 3 retries; accepting result as-is (src/aegis_ai/features/__init__.py:294)
2026-06-17 16:55:47 [WARNING] QualityReview(CVE-2026-4271): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:54 [WARNING] QualityReview(CVE-2026-22822): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:56:12 [WARNING] QualityReview(CVE-2026-22822): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:56:22 [WARNING] QualityReview(CVE-2026-27820): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:56:39 [WARNING] QualityReview(CVE-2026-27820): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:56:41 [WARNING] QualityReview(CVE-2026-22822): output enforcement retry 3/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:56:50 [WARNING] QualityReview(CVE-2026-27820): output enforcement retry 3/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:56:53 [WARNING] QualityReview(CVE-2026-22822): output enforcement exhausted 3 retries; accepting result as-is (src/aegis_ai/features/__init__.py:294)
2026-06-17 16:57:09 [WARNING] QualityReview(CVE-2026-27820): output enforcement exhausted 3 retries; accepting result as-is (src/aegis_ai/features/__init__.py:294)
2026-06-17 17:00:54 [WARNING] QualityReview(CVE-2026-40227): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 17:01:11 [WARNING] QualityReview(CVE-2026-40227): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 17:01:29 [WARNING] QualityReview(CVE-2026-40227): output enforcement retry 3/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 17:01:46 [WARNING] QualityReview(CVE-2026-40227): output enforcement exhausted 3 retries; accepting result as-is (src/aegis_ai/features/__init__.py:294)
2026-06-17 16:52:41 [WARNING] QualityReview(CVE-2023-48795): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:52:43 [WARNING] QualityReview(CVE-2023-53188): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:52:48 [WARNING] QualityReview(CVE-2022-50390): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:53:51 [WARNING] QualityReview(CVE-2025-15284): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:54:26 [WARNING] QualityReview(CVE-2025-39809): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:54:30 [WARNING] QualityReview(CVE-2025-39816): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:54:37 [WARNING] QualityReview(CVE-2025-39822): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:54:44 [WARNING] QualityReview(CVE-2025-39809): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:54:46 [WARNING] QualityReview(CVE-2025-39816): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:54:51 [WARNING] QualityReview(CVE-2025-39822): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:03 [WARNING] QualityReview(CVE-2025-53020): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:18 [WARNING] QualityReview(CVE-2025-53020): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:27 [WARNING] QualityReview(CVE-2025-66623): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:34 [WARNING] QualityReview(CVE-2025-53020): output enforcement retry 3/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:39 [WARNING] QualityReview(CVE-2026-5483): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:44 [WARNING] QualityReview(CVE-2025-53020): output enforcement exhausted 3 retries; accepting result as-is (src/aegis_ai/features/__init__.py:294)
2026-06-17 16:55:47 [WARNING] QualityReview(CVE-2026-4271): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:55:54 [WARNING] QualityReview(CVE-2026-22822): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:56:12 [WARNING] QualityReview(CVE-2026-22822): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:56:22 [WARNING] QualityReview(CVE-2026-27820): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:56:39 [WARNING] QualityReview(CVE-2026-27820): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:56:41 [WARNING] QualityReview(CVE-2026-22822): output enforcement retry 3/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:56:50 [WARNING] QualityReview(CVE-2026-27820): output enforcement retry 3/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 16:56:53 [WARNING] QualityReview(CVE-2026-22822): output enforcement exhausted 3 retries; accepting result as-is (src/aegis_ai/features/__init__.py:294)
2026-06-17 16:57:09 [WARNING] QualityReview(CVE-2026-27820): output enforcement exhausted 3 retries; accepting result as-is (src/aegis_ai/features/__init__.py:294)
2026-06-17 17:00:54 [WARNING] QualityReview(CVE-2026-40227): output enforcement retry 1/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 17:01:11 [WARNING] QualityReview(CVE-2026-40227): output enforcement retry 2/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 17:01:29 [WARNING] QualityReview(CVE-2026-40227): output enforcement retry 3/3: Missing scores for categories: Description — Technical Clarity, Statement — Technical Clarity. You must include criterion scores for all 6 rubric categories. (src/aegis_ai/features/__init__.py:278)
2026-06-17 17:01:46 [WARNING] QualityReview(CVE-2026-40227): output enforcement exhausted 3 retries; accepting result as-is (src/aegis_ai/features/__init__.py:294)
```

Moreover, `QualityReview(CVE-2026-40227)` occasionally times out after 300 seconds when using the [cached fixture](https://github.com/RedHatProductSecurity/aegis-ai/blob/4dd13a56fa9862a9ac506bfe8183a5abf70359fb/evals/osidb_cache/CVE-2026-40227.json) for evals.

**Review by kdudka — COMMENTED** [review]
@gvauter Nice work!  I tried the quality-review feature on a few CVEs and the output seemed decent.  A few remarks inline...

File: src/aegis_ai/features/cve/data_models.py (hunk 5 above)
Comment (gvauter) [reply src/aegis_ai/features/cve/data_models.py:419]: Good catch — added `explanation` field to `QualityReviewModel` so it's consistent with all other `AegisFeatureModel` subclasses. Addressed in the latest push.

File: src/aegis_ai/features/cve/data_models.py (hunk 6 above)
Comment (gvauter) [reply src/aegis_ai/features/cve/data_models.py:469]: Addressed — added explicit rewrite rules to the prompt:

- `suggested_statement` must not duplicate description or mention mitigation/updates
- `suggested_mitigation` must describe config/operational controls only; NEVER suggest updating/upgrading/patching
- If no safe documented mitigation exists, `suggested_mitigation` should be null rather than generic upgrade advice

Verified against CVE-2026-4724 — now suggests disabling JavaScript for untrusted sites instead of 'update to Firefox 149'.

File: src/aegis_ai/features/cve/__init__.py (hunk 7 above)
Comment (gvauter) [reply src/aegis_ai/features/cve/__init__.py:900]: Root cause was the em dash (—) in category names like "Description — Technical Clarity". The LLM was returning regular hyphens (-) which failed the exact string match in `_check_output`.

Fixed by replacing em dashes with hyphens in `CATEGORY_WEIGHTS` (the single source of truth) — this propagates to `_REQUIRED_CATEGORIES`, the prompt text, and all tests.

Verified against all 5 CVEs that were failing (CVE-2023-48795, CVE-2025-53020, CVE-2026-22822, CVE-2026-40227, CVE-2026-4724) — all pass first try with zero retries.

File: src/aegis_ai/features/cve/__init__.py [hunk 8]
```diff
@@ -799,3 +801,384 @@ async def exec(self, cve_id: CVEID, static_context: Any = None):
...
+                "Each criterion must appear exactly once."
+            )
+        missing_criteria = self._REQUIRED_CRITERIA - found_criteria
+        if missing_criteria:
+            return (
+                f"Missing criterion IDs: {', '.join(sorted(missing_criteria))}. "
+                f"You must score all {expected_count} rubric criteria."
+            )
+
+        return None
```
[the remark this answers is not in the document]
Comment (gvauter) [reply src/aegis_ai/features/cve/__init__.py:891]: Skipping for now — the prompt explicitly lists which criteria belong to which category, making this an unlikely LLM failure mode. Adding `_CRITERIA_BY_CATEGORY` would duplicate the mapping a third time (prompt text, `_REQUIRED_CRITERIA`, and the new dict). If we see this in practice we can add it, but haven't observed it in testing.

File: src/aegis_ai/features/cve/data_models.py [hunk 9]
```diff
@@ -346,3 +347,175 @@ class CVSSDiffExplainerModel(AegisFeatureModel):
...
+
+        # Derive rating from weighted score
+        if self.overall_score >= 0.8:
+            self.rating = QualityRating.EXCELLENT
+        elif self.overall_score >= 0.6:
+            self.rating = QualityRating.GOOD
+        elif self.overall_score >= 0.4:
+            self.rating = QualityRating.NEEDS_IMPROVEMENT
+        else:
+            self.rating = QualityRating.FAILS_STANDARDS
```
[the remark this answers is not in the document]
Comment (gvauter) [reply src/aegis_ai/features/cve/data_models.py:516]: Skipping — with our rubric (integer criterion scores 0-2, 5 per category, 6 categories with fixed weights), the edge case of 0.795 rounding to 0.80 requires a very specific score combination. The difference between 0.79 and 0.80 is negligible for a quality tier, and having `overall_score` and `rating` always agree on the stored value is simpler for consumers.

File: tests/test_features.py [hunk 10]
```diff
@@ -139,6 +141,65 @@ async def test_component_intelligence_test_model():
...
+    # Verify overall_score bounds and rating alignment
+    assert 0.0 <= quality_review.overall_score <= 1.0
+    if quality_review.overall_score >= 0.8:
+        assert quality_review.rating == QualityRating.EXCELLENT
+    elif quality_review.overall_score >= 0.6:
+        assert quality_review.rating == QualityRating.GOOD
+    elif quality_review.overall_score >= 0.4:
+        assert quality_review.rating == QualityRating.NEEDS_IMPROVEMENT
+    else:
+        assert quality_review.rating == QualityRating.FAILS_STANDARDS
```
[the remark this answers is not in the document]
Comment (gvauter) [reply tests/test_features.py:183]: Skipping — the model_validator deliberately overrides the LLM's `overall_score` and `rating` with auto-computed values. The raw LLM values being different is expected and by design. We have a dedicated unit test (`test_overall_score_ignores_llm_provided_value`) that verifies this behavior.

File: tests/test_quality_review_models.py [hunk 11]
```diff
@@ -0,0 +1,333 @@
...
+    "criterion_4",
+]
+
+
+def _make_scores(category_scores: list[list[int]]) -> list[CriterionScore]:
+    """Build a flat list of CriterionScore from 6 lists of criterion scores (one per category)."""
+    scores = []
+    for cat_name, cat_scores in zip(_CATEGORY_NAMES, category_scores):
+        for crit_id, s in zip(_CRITERION_IDS, cat_scores):
+            scores.append(
```
[the remark this answers is not in the document]
Comment (gvauter) [reply tests/test_quality_review_models.py:43]: Good catch — added assertions to `_make_scores` to validate input shape before the zip. Addressed in the latest push.

File: src/aegis_ai/features/cve/__init__.py (hunk 7 above)
Comment (kdudka) [reply src/aegis_ai/features/cve/__init__.py:900]: Interesting, I did not expect such an easy fix.

I confirm that these warnings do not appear any more on the same [evaluation set](https://github.com/RedHatProductSecurity/aegis-ai/blob/0dc15ac296c89ddeae4b8a78e1c932715300704e/evals/features/cve/test_suggest_statement.py#L78).

**Review by kdudka — APPROVED** [review]
@gvauter Thanks for the update!  I think this can be safely merged.  Please consider submitting a separate pull request with some evaluation coverage so that we are able to maintain this feature in long term.
